packages/webamp/js/components/Vis.tsx

```diff
@@ -0,0 +1,234 @@
+  useLayoutEffect(() => {
+    analyser.fftSize = 1024;
+  }, [analyser, analyser.fftSize]);
+  const colors = useTypedSelector(Selectors.getSkinColors);
+  let mode = useTypedSelector(Selectors.getVisualizerStyle);
+  const audioStatus = useTypedSelector(Selectors.getMediaStatus);
+  const getWindowShade = useTypedSelector(Selectors.getWindowShade);
+  doubled = useTypedSelector(Selectors.getDoubled);
+  const dummyVizData = useTypedSelector(Selectors.getDummyVizData);
+
+  const dataArray = new Uint8Array(1024);
+  analyser.getByteTimeDomainData(dataArray);
+
+  in_wavedata = new Float32Array(dataArray.length);
+
+  const toggleVisualizerStyle = useActionCreator(Actions.toggleVisualizerStyle);
+  windowShade = getWindowShade("main");
+  // BUG: windowshade does not take into account if the main window is visible (small vis is in pledit)
+  // how can i know the state of individual windows?
+  renderWidth = windowShade ? 38 : 75;
+  renderHeight = windowShade ? 5 : 16;
+  PIXEL_DENSITY = (doubled && windowShade) ? 2 : 1;
+
+  const width = renderWidth * PIXEL_DENSITY;
+  const height = renderHeight * PIXEL_DENSITY;
+
+  const bgCanvas = useMemo(() => {
+    return preRenderBg(
+      width,
+      height,
+      colors[0],
+      colors[1],
+      Boolean(windowShade)
+    );
+  }, [colors, height, width, windowShade]);
+
+  const [canvas, setCanvas] = useState<HTMLCanvasElement | null>(null);
+  // const vis: IVis = {
+  //   canvas,
+  //   analyser
+  // }
+
+  //? painter administration
+  const [painter, setPainter] = useState<VisPaintHandler | null>(null);
+  // const _vis: IVis = useMemo(() => {
+  //   if (!canvas) return { colors, analyser };
+  //   return { canvas, colors, analyser };
+  // }, [analyser, canvas, colors]);
+
+  useEffect(() => {
+    if (!canvas) return;
+    const _setPainter = (PainterType: typeof VisPaintHandler) => {
+      const _vis: IVis = {
+        canvas,
+        colors,
+        analyser,
+        oscStyle: "lines",
+        bandwidth: "wide",
+        coloring: "normal",
+        peaks: true,
+        safalloff: "moderate",
+        sa_peak_falloff: "slow",
+        sa: "analyzer",
+      };
+
+      // uninteruptable painting requires _painter to be always available
+      // const oldPainter = painter;
+      const newPainter = new PainterType(_vis);
+      setPainter(newPainter);
+
+      // not sure it'll achieve the desired effect...
+      // i tried to set the vis mode here, but it didnt quite work out the way i imagined
+/*       switch (_vis.sa) {
+        case "analyzer":
+          mode = VISUALIZERS.BAR;
+          break;
+        case "oscilloscope":
+          mode = VISUALIZERS.OSCILLOSCOPE;
+          // _setPainter(BarPaintHandlerFake);
+          break;
+        case "none":
+          mode = VISUALIZERS.NONE;
+          break;
+        default:
+          mode = VISUALIZERS.NONE;
+      } */
+    };
+    // console.log(" vis mode:", mode);
```

Comment:
Can we remove all commented out code that's just leftover debug code?

Reply:
Yeah, oops :sweat_smile: